README.md
# What is this?

This repositry contains the dotfiles needed to get a working zsh shell in MacOS and setup a Neovim development environment.

# Network install

The network isntall clones the dotfiles project to `~/dotfiles`, in addition to creating links to the dotfiles, and running `mackup restore` for private files not in the repo - such as ssh private keys which are stored in an iCloud folder.

    curl -sS https://raw.githubusercontent.com/yevrah/dotfiles/master/install.sh | /bin/bash




_ptpython/history

# 2018-10-11 09:23:04.916884
+import os; os.sys.path.append('/Users/harvey/dev/mystro.com/website')

# 2018-10-11 09:27:51.302448
+from common.models import *

# 2018-10-11 09:28:41.732297
+users = Individuals.select()

# 2018-10-11 09:28:43.822423
+user

# 2018-10-11 09:28:46.101418
+users

# 2018-10-24 16:08:52.573740
+exit

# 2018-10-24 16:09:27.813810
+from common.models.static import *

# 2018-10-24 16:10:45.466529
+StaticAddBackTypes.select()

# 2018-10-24 16:11:16.704866
+StaticAddressTypes.select()

# 2018-11-21 07:31:52.664347
+exit

# 2018-11-21 07:31:57.528745
+exit()

# 2018-11-21 07:34:19.070620
+import os

# 2018-11-21 07:35:22.487909
+os.path

# 2018-11-21 07:35:34.016267
+os.path.append('./website')

# 2018-11-21 07:36:27.153454
+import sys

# 2018-11-21 07:37:43.251424
+sys.path.append('./website')

# 2018-11-21 07:38:01.451662
+from common.models import *

# 2018-11-21 07:43:25.298513
+sys.modules

# 2018-11-21 07:43:52.179891
+dir(sys)

# 2018-11-21 07:44:18.508478
+sys.modules.items

# 2018-11-21 07:44:24.171237
+sys.modules.keys

# 2018-11-21 07:44:26.876366
+sys.modules.keys()


gdb-test.c
int main(void) {
    int i = 1337;
    return 0;
}


test/test.js
found




found

:C  


setjmp.c
#if 0
set -x

SOURCE=$0
OUT="${SOURCE/\.c/}"

echo "source: $SOURCE, out $OUT"

cc "$SOURCE" -std=c89 -Wall -o "$OUT" && "$OUT" $*
exit
#endif

#include <stdio.h>
#include <setjmp.h>

jmp_buf bufferA, bufferB;

void routineB(); // forward declaration 

void routineA()
{
    int r ;

    printf("(A1)\n");

    r = setjmp(bufferA);
    if (r == 0) routineB();

    printf("(A2) r=%d\n",r);

    r = setjmp(bufferA);
    if (r == 0) longjmp(bufferB, 20001);

    printf("(A3) r=%d\n",r);

    r = setjmp(bufferA);
    if (r == 0) longjmp(bufferB, 20002);

    printf("(A4) r=%d\n",r);
}

void routineB()
{
    int r;

    printf("(B1)\n");

    r = setjmp(bufferB);
    if (r == 0) longjmp(bufferA, 10001);

    printf("(B2) r=%d\n", r);

    r = setjmp(bufferB);
    if (r == 0) longjmp(bufferA, 10002);

    printf("(B3) r=%d\n", r);

    r = setjmp(bufferB);
    if (r == 0) longjmp(bufferA, 10003);
}


int main(int argc, char **argv) 
{
    routineA();
    return 0;
}


ctor_dtor.c
#if 0
set -x

SOURCE=
OUT="${SOURCE/\.c/}"

echo "source: $SOURCE, out $OUT"

cc "$SOURCE" -std=c89 -Wall -o "$OUT" && "$OUT" $*
exit
#endif

include <stdlib.h> /* need malloc and friends */

/* this is the type of object we have, with a single int member */
typedef struct WIDGET_T {
  int member;
} WIDGET_T;

/* functions that deal with WIDGET_T */


/* constructor function */
void
WIDGETctor (WIDGET_T *this, int x)
{
  this->member = x;
}

/* destructor function */
void
WIDGETdtor (WIDGET_T *this)
{
  /* In this case, I really don't have to do anything, but
     if WIDGET_T had internal pointers, the objects they point to
     would be destroyed here.  */
  this->member = 0;
}

/* create function - this function returns a new WIDGET_T */
WIDGET_T *
WIDGETcreate (int m)
{
  WIDGET_T *x = 0;

  x = malloc (sizeof (WIDGET_T));
  if (x == 0) abort (); /* no memory */

  WIDGETctor (x, m);
  return x;
}

/* destroy function - calls the destructor, then frees the object */
void
WIDGETdestroy (WIDGET_T *this)
{
  WIDGETdtor (this);
  free (this);
}

/* END OF CODE */


ptpython/config.py
"""
Configuration for ``ptpython``.
Copy this file to ~/.ptpython/config.py
"""
from __future__ import unicode_literals
from prompt_toolkit.filters import ViInsertMode
from prompt_toolkit.key_binding.input_processor import KeyPress
from prompt_toolkit.keys import Keys
from pygments.token import Token

from ptpython.layout import CompletionVisualisation

__all__ = (
    'configure',
)


def configure(repl):
    """
    Configuration method. This is called during the start-up of ptpython.
    :param repl: `PythonRepl` instance.
    """
    # Show function signature (bool).
    repl.show_signature = True

    # Show docstring (bool).
    repl.show_docstring = True

    # Show the "[Meta+Enter] Execute" message when pressing [Enter] only
    # inserts a newline instead of executing the code.
    repl.show_meta_enter_message = True

    # Show completions. (NONE, POP_UP, MULTI_COLUMN or TOOLBAR)
    repl.completion_visualisation = CompletionVisualisation.POP_UP

    # When CompletionVisualisation.POP_UP has been chosen, use this
    # scroll_offset in the completion menu.
    repl.completion_menu_scroll_offset = 0

    # Show line numbers (when the input contains multiple lines.)
    repl.show_line_numbers = True

    # Show status bar.
    repl.show_status_bar = True

    # When the sidebar is visible, also show the help text.
    repl.show_sidebar_help = True

    # Highlight matching parethesis.
    repl.highlight_matching_parenthesis = True

    # Line wrapping. (Instead of horizontal scrolling.)
    repl.wrap_lines = True

    # Mouse support.
    repl.enable_mouse_support = True

    # Complete while typing. (Don't require tab before the
    # completion menu is shown.)
    repl.complete_while_typing = True

    # Vi mode.
    repl.vi_mode = True

    # Paste mode. (When True, don't insert whitespace after new line.)
    repl.paste_mode = False

    # Use the classic prompt. (Display '>>>' instead of 'In [1]'.)
    repl.prompt_style = 'classic'  # 'classic' or 'ipython'

    # Don't insert a blank line after the output.
    repl.insert_blank_line_after_output = True

    # History Search.
    # When True, going back in history will filter the history on the records
    # starting with the current input. (Like readline.)
    # Note: When enable, please disable the `complete_while_typing` option.
    #       otherwise, when there is a completion available, the arrows will
    #       browse through the available completions instead of the history.
    repl.enable_history_search = False

    # Enable auto suggestions. (Pressing right arrow will complete the input,
    # based on the history.)
    repl.enable_auto_suggest = False

    # Enable open-in-editor. Pressing C-X C-E in emacs mode or 'v' in
    # Vi navigation mode will open the input in the current editor.
    repl.enable_open_in_editor = True

    # Enable system prompt. Pressing meta-! will display the system prompt.
    # Also enables Control-Z suspend.
    repl.enable_system_bindings = True

    # Ask for confirmation on exit.
    repl.confirm_exit = False

    # Enable input validation. (Don't try to execute when the input contains
    # syntax errors.)
    repl.enable_input_validation = True

    # Use this colorscheme for the code.
    #repl.use_code_colorscheme('paraiso-dark')
    repl.use_code_colorscheme('monokai')

    # Enable 24bit True color. (Not all terminals support this. -- maybe check
    # $TERM before changing.)
    repl.true_color = True

    # Install custom colorscheme named 'my-colorscheme' and use it.
    """
    repl.install_ui_colorscheme('my-colorscheme', _custom_ui_colorscheme)
    repl.use_ui_colorscheme('my-colorscheme')
    """

    # Add custom key binding for PDB.
    @repl.add_key_binding(Keys.ControlB)
    def _(event):
        ' Pressing Control-B will insert "pdb.set_trace()" '
        event.cli.current_buffer.insert_text('\nimport pdb; pdb.set_trace()\n')

    # Typing ControlE twice should also execute the current command.
    # (Alternative for Meta-Enter.)
    @repl.add_key_binding(Keys.ControlE, Keys.ControlE)
    def _(event):
        b = event.current_buffer
        if b.accept_action.is_returnable:
            b.accept_action.validate_and_handle(event.cli, b)


    # Typing 'jj' in Vi Insert mode, should send escape. (Go back to navigation
    # mode.)
    @repl.add_key_binding('j', 'j', filter=ViInsertMode())
    def _(event):
        " Map 'jj' to Escape. "
        event.cli.input_processor.feed(KeyPress(Keys.Escape))

    """
    # Custom key binding for some simple autocorrection while typing.
    corrections = {
        'impotr': 'import',
        'pritn': 'print',
    }
    @repl.add_key_binding(' ')
    def _(event):
        ' When a space is pressed. Check & correct word before cursor. '
        b = event.cli.current_buffer
        w = b.document.get_word_before_cursor()
        if w is not None:
            if w in corrections:
                b.delete_before_cursor(count=len(w))
                b.insert_text(corrections[w])
        b.insert_text(' ')
    """


# Custom colorscheme for the UI. See `ptpython/layout.py` and
# `ptpython/style.py` for all possible tokens.
_custom_ui_colorscheme = {
    # Blue prompt.
    Token.Layout.Prompt:                          'bg:#eeeeff #000000 bold',

    # Make the status toolbar red.
    Token.Toolbar.Status:                         'bg:#ff0000 #000000',
}


pythonrc.py
# -*- coding: utf-8 -*-
"""Best goddamn .pythonrc file in the whole world.

This file is executed when the Python interactive shell is started if
$PYTHONSTARTUP is in your environment and points to this file. It's just
regular Python commands, so do what you will. Your ~/.inputrc file can greatly
complement this file.
"""

# Imports we need
import sys
import os
import readline, rlcompleter
import atexit
from pprint import pprint
from tempfile import mkstemp
from code import InteractiveConsole

# Imports we want
import datetime
import pdb

AUTHOR = 'Seth House <seth@eseth.com>'

# Color Support
###############

class TermColors(dict):
    """Gives easy access to ANSI color codes. Attempts to fall back to no color
    for certain TERM values. (Mostly stolen from IPython.)"""

    COLOR_TEMPLATES = (
        ("Black"       , "0;30"),
        ("Red"         , "0;31"),
        ("Green"       , "0;32"),
        ("Brown"       , "0;33"),
        ("Blue"        , "0;34"),
        ("Purple"      , "0;35"),
        ("Cyan"        , "0;36"),
        ("LightGray"   , "0;37"),
        ("DarkGray"    , "1;30"),
        ("LightRed"    , "1;31"),
        ("LightGreen"  , "1;32"),
        ("Yellow"      , "1;33"),
        ("LightBlue"   , "1;34"),
        ("LightPurple" , "1;35"),
        ("LightCyan"   , "1;36"),
        ("White"       , "1;37"),
        ("Normal"      , "0"),
    )

    NoColor = ''
    _base  = '\001\033[%sm\002'

    def __init__(self):
        if os.environ.get('TERM') in ('xterm-color', 'xterm-256color', 'linux',
                                    'screen', 'screen-256color', 'screen-bce'):
            self.update(dict([(k, self._base % v) for k,v in self.COLOR_TEMPLATES]))
        else:
            self.update(dict([(k, self.NoColor) for k,v in self.COLOR_TEMPLATES]))
_c = TermColors()

# Enable AutoComplete
#####################

readline.parse_and_bind('tab:complete')

# Enable a History
##################

HISTFILE="%s/.pyhistory" % os.environ["HOME"]

# Read the existing history if there is one
if os.path.exists(HISTFILE):
    readline.read_history_file(HISTFILE)

# Set maximum number of items that will be written to the history file
readline.set_history_length(300)

def savehist():
    readline.write_history_file(HISTFILE)

atexit.register(savehist)

# Enable Color Prompts
######################

sys.ps1 = '%s>>> %s' % (_c['Green'], _c['Normal'])
sys.ps2 = '%s... %s' % (_c['Red'], _c['Normal'])

# Enable Pretty Printing for stdout
###################################

def my_displayhook(value):
    if value is not None:
        try:
            import __builtin__
            __builtin__._ = value
        except ImportError:
            __builtins__._ = value

        pprint(value)
sys.displayhook = my_displayhook

# Welcome message
#################

WELCOME = """\
%(Cyan)s
You've got color, history, and pretty printing.
(If your ~/.inputrc doesn't suck, you've also
got completion and vi-mode keybindings.)
%(Brown)s
Type \e to get an external editor.
%(Normal)s""" % _c

atexit.register(lambda: sys.stdout.write("""%(DarkGray)s
Sheesh, I thought he'd never leave. Who invited that guy?
%(Normal)s""" % _c))

# Django Helpers
################

def SECRET_KEY():
    "Generates a new SECRET_KEY that can be used in a project settings file."

    from random import choice
    return ''.join(
            [choice('abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)')
                for i in range(50)])

# If we're working with a Django project, set up the environment
# if 'DJANGO_SETTINGS_MODULE' in os.environ:
#     from django.db.models.loading import get_models
#     from django.test.client import Client
#     from django.test.utils import setup_test_environment, teardown_test_environment
#     from django.conf import settings as S
#
#     class DjangoModels(object):
#         """Loop through all the models in INSTALLED_APPS and import them."""
#         def __init__(self):
#             for m in get_models():
#                 setattr(self, m.__name__, m)
#
#     A = DjangoModels()
#     C = Client()
#
#     WELCOME += """%(Green)s
# Django environment detected.
# * Your INSTALLED_APPS models are available as `A`.
# * Your project settings are available as `S`.
# * The Django test client is available as `C`.
# %(Normal)s""" % _c
#
#     setup_test_environment()
#     S.DEBUG_PROPAGATE_EXCEPTIONS = True
#
#     WELCOME += """%(LightPurple)s
# Warning: the Django test environment has been set up; to restore the
# normal environment call `teardown_test_environment()`.
# Warning: DEBUG_PROPAGATE_EXCEPTIONS has been set to True.
# %(Normal)s""" % _c


# Salt Helpers
##############
if 'SALT_CLIENT_CONFIG' in os.environ:
    try:
        import salt.config
        import salt.client
        import salt.runner
    except ImportError:
        pass
    else:
        __opts_client__ = salt.config.client_config(
                os.environ['SALT_CLIENT_CONFIG'])

        # Instantiate LocalClient and RunnerClient
        SLC = salt.client.LocalClient(mopts=__opts_client__)
        SRUN = salt.runner.Runner(__opts_client__)

if 'SALT_MINION_CONFIG' in os.environ:
    try:
        import salt.config
        import salt.client
        import salt.loader
        import jinja2
        import yaml
    except ImportError:
        pass
    else:
        # # Create the Salt __opts__ variable
        __opts__ = salt.config.minion_config(
                os.environ.get('SALT_MINION_CONFIG'))

        # Default to local mode to avoid timeouts if a master is not running.
        # Can set this to 'remote' manually and re-instantiate if desired.
        __opts__['file_client'] = 'local'

        # Instantiate the Caller class
        SCALL = salt.client.Caller(mopts=__opts__)

        # # Populate grains if it hasn't been done already
        if not 'grains' in __opts__ or not __opts__['grains']:
            __opts__['grains'] = salt.loader.grains(__opts__)

        # # Populate template variables
        __salt__ = salt.loader.minion_mods(__opts__)
        __grains__ = __opts__['grains']

# Start an external editor with \e
##################################
# http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/438813/

EDITOR = os.environ.get('EDITOR', 'vim')
EDIT_CMD = '\e'


class EditableBufferInteractiveConsole(InteractiveConsole):
    def __init__(self, *args, **kwargs):
        self.last_buffer = [] # This holds the last executed statement
        InteractiveConsole.__init__(self, *args, **kwargs)

    def runsource(self, source, *args):
        self.last_buffer = [ source.encode('utf-8') ]
        return InteractiveConsole.runsource(self, source, *args)

    def raw_input(self, *args):
        line = InteractiveConsole.raw_input(self, *args)
        if line == EDIT_CMD:
            fd, tmpfl = mkstemp('.py')
            os.write(fd, b'\n'.join(self.last_buffer))
            os.close(fd)
            os.system('%s %s' % (EDITOR, tmpfl))
            line = open(tmpfl).read()
            os.unlink(tmpfl)
            tmpfl = ''
            lines = line.split( '\n' )
            for i in range(len(lines) - 1): self.push( lines[i] )
            line = lines[-1]
        return line


c = EditableBufferInteractiveConsole(locals=locals())
c.interact(banner=WELCOME)

# Exit the Python shell on exiting the InteractiveConsole
sys.exit()
